File: frontend/src/api/momentum.ts
import { useMutation } from '@tanstack/react-query';
import { apiFetch } from './client';

export interface MacdResult {
  macd: number;
  signal: number;
  hist: number;
  prevMacd: number;
  prevSig: number;
}

export interface BollingerBands {
  upper: number;
  mid: number;
  lower: number;
  bw: number;
}

export interface MomentumScoreBreakdown {
  rsi: number;
  macd: number;
  volume: number;
  trend: number;
  riskReward: number;
  total: number;
}

export type MomentumSignal = 'STRONG BUY' | 'BUY' | 'WATCH' | 'AVOID';

export interface MomentumAnalysis {
  price: number;
  sma20: number;
  sma50: number;
  rsi: number;
  macd: MacdResult;
  bb: BollingerBands;
  volRatio: number;
  dayChg: number;
  score: MomentumScoreBreakdown;
  signal: MomentumSignal;
  entryLow: number;
  entryHigh: number;
  entryMid: number;
  stopLoss: number;
  target: number;
  rr: number;
  flags: string[];
  extras: string[];
}

export interface MomentumResponse {
  symbol: string;
  name: string | null;
  analysis: MomentumAnalysis;
}

// A mutation, not a query — this is triggered on-demand by a ticker lookup
// (button press), not something to auto-refetch/cache by symbol.
export function useMomentumAnalysis() {
  return useMutation({
    mutationFn: (symbol: string) => apiFetch<MomentumResponse>(`/momentum/${encodeURIComponent(symbol)}`),
  });
}

File: analysis-service/app/models/momentum.py
from typing import Literal

from pydantic import BaseModel

# Mirrors backend/src/services/momentum.service.ts's types field-for-field
# (same hand-maintained-duplicate caveat as Long-Term Analysis/Contrarian
# Comeback). calcKellySizing has no Python/request-model equivalent here —
# it stays client-side (frontend/src/lib/kelly.ts), never round-trips.


class MomentumAnalysisRequest(BaseModel):
    closes: list[float]
    lows: list[float]
    volumes: list[float]
    price: float


class MacdResult(BaseModel):
    macd: float
    signal: float
    hist: float
    prevMacd: float
    prevSig: float


class BollingerBands(BaseModel):
    upper: float
    mid: float
    lower: float
    bw: float


class MomentumScoreBreakdown(BaseModel):
    rsi: int
    macd: int
    volume: int
    trend: int
    riskReward: int
    total: int


class MomentumAnalysisResponse(BaseModel):
    price: float
    sma20: float
    sma50: float
    rsi: float
    macd: MacdResult
    bb: BollingerBands
    volRatio: float
    dayChg: float
    score: MomentumScoreBreakdown
    signal: Literal['STRONG BUY', 'BUY', 'WATCH', 'AVOID']
    entryLow: float
    entryHigh: float
    entryMid: float
    stopLoss: float
    target: float
    rr: float
    flags: list[str]
    extras: list[str]
